r-package/grf/vignettes/rate_cv.Rmd

```diff
@@ -0,0 +1,438 @@
+text(4.5, 2.25, offset = 0, "4", col = "red")
+
+lines(c(1, 5), c(1,1), lwd = 4, col = "blue")
+lines(c(5, 6), c(1,1), lwd = 4, col = "red")
+text(1.5, 1.25, offset = 0, "1", col = "blue")
+text(2.5, 1.25, offset = 0, "2", col = "blue")
+text(3.5, 1.25, offset = 0, "3", col = "blue")
+text(4.5, 1.25, offset = 0, "4", col = "blue")
+text(5.5, 1.25, offset = 0, "5", col = "red")
+
+text(3, 5.5, "Train", col = "blue")
+text(4, 5.5, "Test", col = "red")
+```
+
+This construction is going to make our $t$-values satisfy a martingale property and allow for valid aggregation. The steps are:
+
+1. Randomly split the data into $k=1,\ldots,K$ evenly sized folds.
+
+2. For each fold $k=2,\ldots,K$:
+
+    Obtain an estimated CATE function using $\color{blue}{\text{training}}$ data from folds $1,\ldots,k-1$.
+
+    Form a RATE estimate using data from the $k$-th $\color{red}{\text{test}}$ fold and compute a $t$-statistic.
+
+3. Aggregate the $K-1$ different $t$-statistics and use this as a final test statistic.
+
+This procedure is coded in the simple function below.
+```{r}
+# Estimate RATE on K-1 random sequential folds to get K-1 t-statistics, then return a p-value
+# for H0 using the sum of these t-values
+rate_sequential = function(X, Y, W, num.folds = 5) {
+  fold.id = sample(rep(1:num.folds, length = nrow(X)))
+  samples.by.fold = split(seq_along(fold.id), fold.id)
+
+  t.statistics = c()
+
+  # Form AIPW scores for estimating RATE
+  nuisance.forest = causal_forest(X, Y, W)
+  DR.scores = get_scores(nuisance.forest)
+
+  for (k in 2:num.folds) {
+    train = unlist(samples.by.fold[1:(k - 1)])
+    test = samples.by.fold[[k]]
+
+    cate.forest = causal_forest(X[train, ], Y[train], W[train])
+
+    cate.hat.test = predict(cate.forest, X[test, ])$predictions
+
+    rate.fold = rank_average_treatment_effect.fit(DR.scores[test], cate.hat.test)
+    t.statistics = c(t.statistics, rate.fold$estimate / rate.fold$std.err)
+  }
+
+  p.value = 2 * pnorm(-abs(sum(t.statistics) / sqrt(num.folds - 1)))
+
+  p.value
+}
+```
+
+If we repeat the check on data with no heterogeneity we can see this construction gives us the correct rejection rate (within the error tolerance of our Monte Carlo repetitions):
+```{r, eval=FALSE}
+res = t(replicate(500, {
+  n = 1500
+  p = 5
+  X = matrix(rnorm(n * p), n, p)
+  W = rbinom(n, 1, 0.5)
+  Y = pmin(X[, 3], 0) + rnorm(n)
+
+  pval.sequential = rate_sequential(X, Y, W)
+
+  c(
+    reject.sequential = pval.sequential <= 0.05
+  )
+}))
+paste("Rejection rate:", mean(res) * 100, "%")
+#> [1] "Rejection rate: 5.4 %"
+```
+
+Now that we have a valid test for treatment heterogeneity using several folds, let's try it out on a simulated example and compare its performance with simple sample splitting. Below we generate data from synthetic observational data (Setup A in Nie & Wager (2021)) where treatment effect heterogeneity is hard to detect. In this case, we want our test to reject $H_0$ of no heterogeneity.
```

Comment:
Make it explicit that you're also implementing test/train sample splitting as a baseline; that test/train splitting is what's proven to be valid in the RATE paper; but that this is potentially lower power since you only have 50% of the data for validation.